### ISSUE TITLE
SSH Negotiation Failure Against Gerrit 2.14.2

### ISSUE THREAD

**Issue by markelkins** [MeetMe/jira-gerrit-plugin#51]
I recently upgraded my Gerrit server to 2.14.2 and now the plugin is failing to connect complaining about an SSH algorithm negotiation failure. Based on what I'm reading online, the most likely culprit is that the JSch library being used is out of date and only supports old key agreement algorithms. The exception generated is below:

```
com.sonyericsson.hudson.plugins.gerrit.gerritevents.ssh.SshException: com.jcraft.jsch.JSchException: Algorithm negotiation fail
	at com.sonyericsson.hudson.plugins.gerrit.gerritevents.ssh.SshConnectionImpl.<init>(SshConnectionImpl.java:126)
	at com.sonyericsson.hudson.plugins.gerrit.gerritevents.ssh.SshConnectionFactory.getConnection(SshConnectionFactory.java:76)
	at com.sonyericsson.hudson.plugins.gerrit.gerritevents.GerritQueryHandler.runQuery(GerritQueryHandler.java:243)
	at com.sonyericsson.hudson.plugins.gerrit.gerritevents.GerritQueryHandler.queryJava(GerritQueryHandler.java:132)
	at com.meetme.plugins.jira.gerrit.adminui.AdminServlet.performConnectionTest(AdminServlet.java:189)
	at com.meetme.plugins.jira.gerrit.adminui.AdminServlet.handlePost(AdminServlet.java:171)
	at com.meetme.plugins.jira.gerrit.adminui.AdminServlet.doPost(AdminServlet.java:121)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:648)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	at com.atlassian.plugin.servlet.DelegatingPluginServlet.service(DelegatingPluginServlet.java:37)
...
```

From what I can tell, updating this plugin to the latest version of [gerrit-events](https://github.com/sonyxperiadev/gerrit-events) would update Jsch to 0.1.51, which supports the newer key agreement algorithms.

**Comment by conz27** [comment]
I just tried bumping gerrit-events to 1.99.1 (latest), and I was able to compile. I confirmed that 1.99.1 comes with JSch v0.1.50, not v0.1.51.

@markelkins, do you know which key-agreement protocol you need?

**Comment by mf69** [comment]
Unable to reproduce.
Expected to see this issue on gerrit 2.14.2 <-> jira 7.4.2 w/ Jira Gerrit plugin 3.2
but instead it works.

**Comment by mf69** [comment]
@markelkins now I'm curious what version of Jira and the plugin **you** have, and other details like the underlying Java VM.

**Comment by markelkins** [comment]
@mf69 I am running JIRA 7.2.6 with version 3.2 of the plugin. The JVM is Oracle 1.8.0_47-b02. I don't expect that the JIRA install is the issue as everything was working before we jumped to Gerrit 2.14.2. For security reasons we do restrict the SSH ciphers and algorithms available of our Gerrit server. For reference, here is our config:

```
[sshd]
    listenAddress = *:29418
    cipher = -3des-cbc
    cipher = -blowfish-cbc
    cipher = -none
    kex = -diffie-hellman-group-exchange-sha1
    kex = -diffie-hellman-group14-sha1
    kex = -diffie-hellman-group1-sha1
    mac = -hmac-md5
    mac = -hmac-md5-96
    mac = -hmac-sha1
    mac = -hmac-sha1-96
```

**Comment by mf69** [comment]
FWIW we make no sshd restrictions on the Gerrit side and after upgrading to 2.14.2 we have observed no problems with the Jira plugin as-is.

**Comment by markelkins** [comment]
@mf69 I suspect that is true. The SSH algorithms we are disabling are all considered broken and deprecated. Nobody should still be using them at this point. I'm not sure why Gerrit's SSHD allowed them in 2.13.x even through we had them disabled in the config, but 2.14.x is definitely rejecting them.

I posted a pull request that upgrades the gerrit-events dependency to 1.99.1 (newest 1.x version that I could find), which upgrades Jsch to 0.1.50. Unfortunately to get support for modern SSH ciphers Jsch has to be at least 0.1.51, so I also explicitly pulled in 0.1.54 (the latest) as a dependency. This fixed the issues negotiating SSH connections from the plugin to our Gerrit server. So far no issues running on our production servers for the last week.

@jhansche For security reasons, I would recommend you seriously consider accepting the pull request. Sticking with the deprecated algorithms makes the plugin an attractive attack vector for someone looking to steal data from a Gerrit server.

**Comment by jhansche** [comment]
@markelkins Thanks, I've merged the PR and will look at releasing an update shortly

**Comment by jhansche** [comment]
This has been released in v3.3:
https://marketplace.atlassian.com/plugins/com.meetme.plugins.jira.gerrit-plugin/versions#b77
https://github.com/MeetMe/jira-gerrit-plugin/releases/tag/gerrit-plugin-3.3
